bdk_chain/src/keychain/keychain_txout_index.rs

```diff
@@ -195,92 +284,130 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+    /// Get the last derivation index revealed for `keychain`.
+    pub fn last_revealed_index(&self, keychain: &K) -> Option<u32> {
+        self.last_revealed.get(keychain).cloned()
     }
 
-    /// Convenience method to call [`Self::store_up_to`] on several keychains.
-    pub fn store_all_up_to(&mut self, keychains: &BTreeMap<K, u32>) -> DerivationAdditions<K> {
+    /// Convenience method to call [`Self::reveal_to_target`] on multiple keychains.
+    pub fn reveal_to_target_multi(
+        &mut self,
+        keychains: &BTreeMap<K, u32>,
+    ) -> (
+        BTreeMap<K, impl Iterator<Item = (u32, Script)>>,
+        DerivationAdditions<K>,
+    ) {
         let mut additions = DerivationAdditions::default();
+        let mut spks = BTreeMap::new();
+
         for (keychain, &index) in keychains {
-            additions.append(self.store_up_to(keychain, index));
+            let (new_spks, new_additions) = self.reveal_to_target(&keychain, index);
+            if let Some(new_spks) = new_spks {
+                spks.insert(keychain.clone(), new_spks);
+                additions.append(new_additions);
+            }
         }
-        additions
+
+        (spks, additions)
     }
 
-    /// Derives script pubkeys from the descriptor **up to and including** `up_to` and stores them
-    /// (if necessary).
-    ///
-    /// Returns [`DerivationAdditions`] for any new `script_pubkey`s that has been added. If no
-    /// script pubkeys are added, or if `keychain` does not exist, [`DerivationAdditions`] will be
-    /// empty.
-    pub fn store_up_to(&mut self, keychain: &K, up_to: u32) -> DerivationAdditions<K> {
-        let descriptor = match self.keychains.get(&keychain) {
-            Some(descriptor) => descriptor,
-            None => return Default::default(),
-        };
-
-        let end = match descriptor.has_wildcard() {
-            false => 0,
-            true => up_to,
-        };
-        let (next_to_derive, can_derive) = self.next_derivation_index(keychain);
-        if !can_derive || next_to_derive > end {
-            return Default::default();
+    /// Reveals script pubkeys of the `keychain`'s descriptor **up to and including** the
+    /// `target_index`.
+    ///
+    /// If the `target_index` cannot be reached (due to the descriptor having no wildcard, and/or
+    /// the `target_index` is in the hardened index range), this method will do a best-effort and
+    /// reveal up to the last possible index.
+    ///
+    /// This returns an iterator of newly revealed indices (along side their scripts), and a
+    /// [`DerivationAdditions`] which reports updates to the latest revealed index. If no new script
+    /// pubkeys are revealed, both of these will be empty.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `keychain` does not exist.
+    pub fn reveal_to_target(
+        &mut self,
+        keychain: &K,
+        target_index: u32,
+    ) -> (
+        Option<impl Iterator<Item = (u32, Script)>>,
```

Comment:
What is the semantics of `None` here? If it just means the iterator would be empty just return an empty iterator.

Reply:
Fixed in 7e664feabc3c37919bb774d192a8864c1aaa7843